src/collision/cm_trace.c

```diff
@@ -414,130 +371,112 @@ next:;
 
 	const float midf2 = p1f + (p2f - p1f) * frac2;
 
-	if (midf2 < data.unnudged_fraction) {
+	if (midf2 < data->unnudged_fraction) {
 		const vec3_t mid = Vec3_Mix(p1, p2, frac2);
 		
 		num = node->children[side ^ 1];
 
 		// if < 0, we are in a leaf node
 		if (num < 0) {
-			Cm_TraceToLeaf(-1 - num);
+			Cm_TraceToLeaf(data, -1 - num);
 		} else {
-			Cm_TraceToNode(num, midf2, p2f, mid, p2);
+			Cm_TraceToNode(data, num, midf2, p2f, mid, p2);
 		}
 	}
 }
 
 
 /**
- * @brief Primary collision detection entry point. This function recurses down
- * the BSP tree from the specified head node, clipping the desired movement to
- * brushes that match the specified contents mask.
- *
- * @param start The starting point.
- * @param end The desired end point.
- * @param bounds The bounding box, in model space.
- * @param head_node The BSP head node to recurse down. For inline BSP models,
- * the head node is the root of the model's subtree. For mesh models, a
- * special reserved box hull and head node are used.
- * @param contents The contents mask to clip to.
- * @param matrix The matrix to adjust tested planes by.
- *
- * @return The trace.
+ * @brief
  */
-static inline cm_trace_t Cm_BoxTrace_(void) {
+static inline void Cm_BoxTrace_(cm_trace_data_t *data) {
 
-	if (!cm_bsp.num_nodes) { // map not loaded
-		return data.trace;
+	if (!cm_bsp.num_nodes) {
+		return;
 	}
 
-	Box3_ToPoints(data.bounds, data.offsets);
-
-	memset(data.brush_cache, 0xff, sizeof(data.brush_cache));
-
 	// check for position test special case
-	if (Vec3_Equal(data.start, data.end)) {
+	if (Vec3_Equal(data->start, data->end)) {
 		static __thread int32_t leafs[MAX_BSP_LEAFS];
-		box3_t abs_bounds = data.abs_bounds;
 
-		if (data.is_transformed) {
-			abs_bounds = Mat4_TransformBounds(data.inverse_matrix, abs_bounds);
-		}
-
-		const size_t num_leafs = Cm_BoxLeafnums(abs_bounds,
+		const size_t num_leafs = Cm_BoxLeafnums(data->abs_bounds,
 												leafs,
 												lengthof(leafs),
 												NULL,
-												data.head_node);
+												data->head_node);
 
 		for (size_t i = 0; i < num_leafs; i++) {
-			Cm_TestInLeaf(leafs[i]);
+			Cm_TestInLeaf(data, leafs[i]);
 
-			if (data.trace.all_solid) {
+			if (data->trace.all_solid) {
 				break;
 			}
 		}
 
-		data.trace.end = data.start;
-		return data.trace;
+		data->trace.end = data->start;
+		return;
 	}
 
-	if (data.is_transformed) {
-		data.size = Box3_Symetrical(Mat4_TransformBounds(data.inverse_matrix, Box3_Expand(data.bounds, BOX_EPSILON)));
-		Cm_TraceToNode(data.head_node, 0.f, 1.f, Mat4_Transform(data.inverse_matrix, data.start), Mat4_Transform(data.inverse_matrix, data.end));
-	} else {
-		data.size = Box3_Symetrical(Box3_Expand(data.bounds, BOX_EPSILON));
-		Cm_TraceToNode(data.head_node, 0.f, 1.f, data.start, data.end);
-	}
+	Cm_TraceToNode(data, data->head_node, 0.f, 1.f, data->start, data->end);
```

Comment:
Similarly, all of these transforms, if required, are done in Cm_TransformedBoxTrace.